content/post/hercules.v4.md

```diff
@@ -0,0 +1,295 @@
+{{% /caption %}}
+
+Reproduce:
+```
+hercules --couples --pb https://github.com/tensorflow/tensorflow | labours.py -m couples -o tensorflow
+```
+
+We can repeat the same trick with the developers: they are couples if they commit
+to the same files. [Interactive example on Tensorflow developers.](http://projector.tensorflow.org/?config=http://vmarkovtsev.github.io/projector/tf_people.json)
+
+Finally, the trick works once again with structural units in the source code. We parse
+each file with [Babelfish](https://doc.bblf.sh) and extract anything we wish: functions,
+classes, etc. Then we apply the same coupling analysis to them: we know which are changed
+in each commit.
+
+By default, the functions are extracted.
+```
+hercules --shotness --pb https://github.com/tensorflow/tensorflow | labours.py -m shotness -o tensorflow
+```
+
+#### Comment sentiment
+
+This is a proof-of-concept for running a Tensorflow model over the source code.
+We take [BiDiSentiment](https://github.com/vmarkovtsev/BiDiSentiment) and apply it to comments.
+That model is general-purpose and the result is often weird, but it works.
+I wrote about it in [the other blog post](../codesent).
+
+Your binary must be compiled with Tensorflow support (the released one is **not**).
+```
+hercules --sentiment --pb https://github.com/tensorflow/tensorflow | labours.py -m sentiment
+```
+
+# v4 architecture
+
+Hercules is built on top of the `Pipeline` which is a Directed Acyclic Graph (DAG) of `PipelineItem`-s.
+Each pipeline item has a name and specifies its dependencies and what it provides in return.
+For example, a `RenameAnalysis` item detects the renamed and slightly changed files between two commits.
+It consumes the list of changes and the list of corresponding blobs. It calculates
+a better list of changes with some of the "delete + create" pairs replaced with
+single "edit under new name" elements. Yep, Git packfiles do not store this kind
+of information (intentionally) and `git` detects renames using heuristics every
+time you execute it.
+
+Hercules builds the DAG of pipeline items automatically based on their dependencies
+and provided intermediate results. Some items try to improve on the others -
+`RenameAnalysis` for example, and Hercules understands that and correctly resolves
+the structure: all the rest of the items which depend on that type of analysis are
+pointed from the "improved" ones, and the "improved" ones in turn depend on the
+"basic" ones. Here is an example of the built pipeline to calculate the burndown
+stats:
+
+{{% caption src="/post/hercules.v4/pipeline.png" %}}
+Pipeline DAG example.
+{{% /caption %}}
+
+Once the pipeline is built, each item changes its state in certain ways.
+Here is the most full example of the "leaf" item which does not provide any
```

Comment:
I'm not sure I understand this sentence. What are those "certain ways"? Either drop it or explain it.
Also in what way this is the most full example? Maybe replace "the most full example" by "a complete example" or "an end-to-end example".